python/nistoar/midas/dbio/base.py

```diff
@@ -941,21 +943,53 @@ def user_groups(self) -> frozenset:
             self.recache_user_groups()
         return self._whogrps
 
+    def all_groups_for(self, who) -> frozenset:
+        """
+        Return the frozen set of all groups a user or group belongs to.
+        """
+        adhoc = self.groups.select_ids_for_user(who)
+        virtual_groups = self._get_virtual_groups_for(who)
+        all_groups = frozenset(adhoc.union(virtual_groups))
+
+        return all_groups
+
     @property
     def people_service(self) -> PeopleService:
         """
-        an attached PeopleService instance or None if such a service is not available.  This service 
+        an attached PeopleService instance or None if such a service is not available.  This service
         encapsulates access to the organization's staff directory service.
         """
         return self._peopsvc
 
     def recache_user_groups(self):
         """
-        the :py:property:`user_groups` contains a cached list of all the groups the user is 
-        a member of.  This function will recache this list (resulting in queries to the backend 
-        database).  
-        """
-        self._whogrps = frozenset(self.groups.select_ids_for_user(self._who))
+        the :py:property:`user_groups` contains a cached list of all the groups the user is
+        a member of.  This function will recache this list (resulting in queries to the backend
+        database).
+        """
+        adhoc_groups = self.groups.select_ids_for_user(self._who)
+        virtual_groups = self._get_virtual_groups_for(self._who)
+        self._whogrps = frozenset(adhoc_groups.union(virtual_groups))
+
+    def _get_virtual_groups_for(self, user_id: str) -> List[str]:
+        """
+        Return the list of 'virtual groups' ids a user is part of, based on the staff directory
+        (PeopleService). Returns an empty list if the PeopleService is not available or the
+        user is not found.
+        """
+        if not self.people_service:
+            return []
+        person = self.people_service.get_person_by_eid(user_id)
+        if not person:
+            return []
+        out = []
+        if 'ouNumber' in person and person['ouNumber']:
+            out.append(f"ouNumber:{person['ouNumber']}")
+        if 'divisionNumber' in person and person['divisionNumber']:
+            out.append(f"divisionNumber:{person['divisionNumber']}")
+        if 'groupNumber' in person and person['groupNumber']:
+            out.append(f"groupNumber:{person['groupNumber']}")
+        return out
```

Comment:
Let's change the group prefix from `ouNumber`, `divisionNumber`, and `groupNumber` to `nistou`, `nistdiv`, `nistgrp`, respectively.